Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef CartesianPowerIterator<InputIt> const_iterator;
+
+    CartesianPower(InputIt first_, InputIt last_, size_t power_);
+    template<typename Container>
+    CartesianPower(const Container& container, size_t power_);
+
+    const_iterator cbegin() const;
+    const_iterator cend() const;
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    size_t size() const;
+    std::pair<size_t, size_t> size_2d(size_t resolution) const;
+
+    InputIt first;
+    InputIt last;
```

Comment:
`first` and `last` can be any pair of iterators that define a valid range (that is, `last` should be reachable from `first`). They needn't be the `begin()` and `end()` of a container.

Changing `first`, `last`, or `power` will not invalidate any associated  `CartesianPowerIterator`s, but could allow `CartesianPower` instances to be reused or initialised outside the constructor. I don't think it does any harm to make them non-const, and future code may need to change them.